### PR TITLE
8265917: Different values computed by C2 and interpreter/C1 for Math.pow(x, 2.0) on x86_32

### PR DESCRIPTION
**Pull request by DamonFool**
Hi all,

C2 may produce different results for Math.pow(x, 2.0) compared with interpreter/C1 on x86_32.

E.g., for Math.pow(1.0 / 2047, 2.0):
```
interpreter: 2.38651580386563E-7 
         C2: 2.3865158038656307E-7
```

The reason is that C2 will replace Math.pow(x, 2.0) with x*x.
However, there is no such optimization for interpreter/C1 on x86_32.

The fix just disables C2's opt for Math.pow(x, 2.0) on x86_32 since nobody (or very few people) would run x86_32.
And we don't have a plan to implement such opt on x86_32.

Another reason to fix this bug is that we need this patch to extend C2's opt for Math.pow(x, 0.5) on other platforms.

Thanks.
Best regards,
Jie

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8265917](https://bugs.openjdk.java.net/browse/JDK-8265917): Different values computed by C2 and interpreter/C1 for Math.pow(x, 2.0) on x86_32

### Reviewers
 * [Vladimir Kozlov](https://openjdk.java.net/census#kvn) (@vnkozlov - **Reviewer**)
 * [Tobias Hartmann](https://openjdk.java.net/census#thartmann) (@TobiHartmann - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/3677/head:pull/3677` \
`$ git checkout pull/3677`

Update a local copy of the PR: \
`$ git checkout pull/3677` \
`$ git pull https://git.openjdk.java.net/jdk pull/3677/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 3677`

View PR using the GUI difftool: \
`$ git pr show -t 3677`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/3677.diff">https://git.openjdk.java.net/jdk/pull/3677.diff</a>

</details>

### CODE REVIEW

**Comment by vnkozlov** [comment]
I think you should fix 32-bit version of `dpow` stub because 64-bit stub has this optimization:
https://github.com/openjdk/jdk/blob/master/src/hotspot/cpu/x86/macroAssembler_x86_pow.cpp#L828

Also add `pow(x, 0.5)` as well.

C `dpow` code in VM has this optimization which works in 32- and in 64-bit when stub is not used (for example, on other platforms):
https://github.com/openjdk/jdk/blob/master/src/hotspot/share/runtime/sharedRuntimeTrans.cpp#L496

Java lib code also has this optimization:
https://github.com/openjdk/jdk/blob/master/src/java.base/share/classes/java/lang/FdLibm.java#L362

Only 32-bit x86 stub is missing it:
https://github.com/openjdk/jdk/blob/master/src/hotspot/cpu/x86/macroAssembler_x86_pow.cpp#L2544

The stub is generated only on x86:
```
$ grep -r StubRoutines::_dpow src/hotspot/cpu/
src/hotspot/cpu//x86/stubGenerator_x86_64.cpp:        StubRoutines::_dpow = generate_libmPow();
src/hotspot/cpu//x86/stubGenerator_x86_32.cpp:        StubRoutines::_dpow = generate_libmPow();
```
I suggest to fix it to be consistent with the rest of VM and Java lib code.

**Comment by DamonFool** [comment]
> I think you should fix 32-bit version of `dpow` stub because 64-bit stub has this optimization:
> https://github.com/openjdk/jdk/blob/master/src/hotspot/cpu/x86/macroAssembler_x86_pow.cpp#L828
> 
> Also add `pow(x, 0.5)` as well.
> 
> C `dpow` code in VM has this optimization which works in 32- and in 64-bit when stub is not used (for example, on other platforms):
> https://github.com/openjdk/jdk/blob/master/src/hotspot/share/runtime/sharedRuntimeTrans.cpp#L496
> 
> Java lib code also has this optimization:
> https://github.com/openjdk/jdk/blob/master/src/java.base/share/classes/java/lang/FdLibm.java#L362
> 
> Only 32-bit x86 stub is missing it:
> https://github.com/openjdk/jdk/blob/master/src/hotspot/cpu/x86/macroAssembler_x86_pow.cpp#L2544
> 
> The stub is generated only on x86:
> 
> ```
> $ grep -r StubRoutines::_dpow src/hotspot/cpu/
> src/hotspot/cpu//x86/stubGenerator_x86_64.cpp:        StubRoutines::_dpow = generate_libmPow();
> src/hotspot/cpu//x86/stubGenerator_x86_32.cpp:        StubRoutines::_dpow = generate_libmPow();
> ```
> 
> I suggest to fix it to be consistent with the rest of VM and Java lib code.

Hi @vnkozlov ,

Math.pow(x, 2.0) has been optimized for x86_32.
As for the Math.pow(x, 0.5), it will be handled in JDK-8265940 after this issue.

Thanks.

**Comment by DamonFool** [comment]
Testing:
  -tier1 ~ tier3 on Linux/x86_32, no regression

**Review by vnkozlov — APPROVED** [review]
Good.

**Comment by DamonFool** [comment]
> Good.

Thanks @vnkozlov for your review.

Let's wait for a second review of this change.
Thanks.

File: src/hotspot/cpu/x86/macroAssembler_x86_pow.cpp
Comment (TobiHartmann) [inline src/hotspot/cpu/x86/macroAssembler_x86_pow.cpp:2546]: Constants in the 64-bit code and `_static_const_table_pow` above use `ATTRIBUTE_ALIGNED`. Should we add that for `DOUBLE2` as well?

File: src/hotspot/cpu/x86/macroAssembler_x86_pow.cpp
Comment (TobiHartmann) [inline src/hotspot/cpu/x86/macroAssembler_x86_pow.cpp:2555]: 64-bit code also handles `Special case: pow(x, 0.5) => sqrt(x)`. Don't we need to add that as well for consistency?

File: src/hotspot/cpu/x86/macroAssembler_x86_pow.cpp
Comment (DamonFool) [reply src/hotspot/cpu/x86/macroAssembler_x86_pow.cpp]: Hi @TobiHartmann ,

Thanks for your review.

Yes, it seem better to be with ATTRIBUTE_ALIGNED.
Updated.

File: src/hotspot/cpu/x86/macroAssembler_x86_pow.cpp
Comment (DamonFool) [reply src/hotspot/cpu/x86/macroAssembler_x86_pow.cpp:2557]: > 64-bit code also handles `Special case: pow(x, 0.5) => sqrt(x)`. Don't we need to add that as well for consistency?

This is the bug fix for C2's optimization of pow(x, 2.0).

pow(x, 0.5) will be handled in JDK-8265940 .
Thanks.

**Review by TobiHartmann — APPROVED** [review]
Looks good to me.

**Comment by DamonFool** [comment]
> Looks good to me.

Thanks @TobiHartmann .

@vnkozlov , are you OK with the latest change?
Thanks.

**Review by vnkozlov — APPROVED** [review]
Latest version with alignment is good.

**Comment by DamonFool** [comment]
> Latest version with alignment is good.

Thanks @vnkozlov .
/integrate

**Pull request closed without merging**
